spark/spark-version/2.0/src/main/scala/org/apache/spark/storage/BlockManagerWrapper.scala

```diff
@@ -61,6 +66,35 @@ object BlockManagerWrapper {
     }
   }
 
+  private val getLocalBytesFn: (BlockId) => Option[ByteBuffer] = {
+    val bmClass = classOf[BlockManager]
+    val getLocalBytesMethod = bmClass.getMethod("getLocalBytes", classOf[BlockId])
+
+    // Spark versions before 2.2.0 declare:
+    // def getLocalBytes(blockId: BlockId): Option[ChunkedByteBuffer]
+    // Spark 2.2.0+ declares:
+    // def getLocalBytes(blockId: BlockId): Option[BlockData]
+    // Because the latter change happened in the commit that introduced BlockData,
+    // and because you can't discover the generic type of the return type by reflection,
+    // distinguish the cases by seeing if BlockData exists.
+    try {
+      val blockDataClass = Class.forName("org.apache.spark.storage.BlockData")
+      // newer method, apply reflection to transform BlockData after invoking
+      val toByteBufferMethod = blockDataClass.getMethod("toByteBuffer")
+      (blockId: BlockId) =>
+        getLocalBytesMethod.invoke(SparkEnv.get.blockManager, blockId)
+          .asInstanceOf[Option[_]]
+          .map(blockData => toByteBufferMethod.invoke(blockData).asInstanceOf[ByteBuffer])
+    } catch {
+      case _: ClassNotFoundException =>
+        // older method, can be invoked directly
+        (blockId: BlockId) =>
+          getLocalBytesMethod.invoke(SparkEnv.get.blockManager, blockId)
+            .asInstanceOf[Option[ChunkedByteBuffer]]
+            .map(_.toByteBuffer)
+    }
+  }
+
```

Comment:
getLocalBytesFn will be heavily called during training. In current implementation with spark 2.0, spark 2.1, code will frequently go to "catch exception" block and it will affect training performance. Can we add a flag which is used to mark if BlockData exists and set it for the first time getLocalBytesFn called. After that we can simply check this flag to decide which getLocalBytesMethod should be invoked. In this way, we don't need go to catch execption every time getLocalBytesFn is called which can avoid performance downgrade.